pkg/apis/ceph.rook.io/v1/types.go

```diff
@@ -1588,6 +1588,185 @@ type ObjectZoneSpec struct {
+	// The ARN of the topic generated by the RGW
+	// +optional
+	// +nullable
+	ARN *string `json:"ARN,omitempty"`
+}
+
+// CephBucketTopicList represents a list Ceph Object Store Bucket Notification Topics
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type CephBucketTopicList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []CephBucketTopic `json:"items"`
+}
+
+// BucketTopicSpec represent the spec of a Bucket Topic
+type BucketTopicSpec struct {
+	// The name of the object store on which to define the topic
+	// +kubebuilder:validation:MinLength=1
+	ObjectStoreName string `json:"objectStoreName"`
+	// The namespace of the object store on which to define the topic
+	// +kubebuilder:validation:MinLength=1
+	ObjectStoreNamespace string `json:"objectStoreNamespace"`
+	// Data which is sent in each event
+	// +optional
+	OpaqueData string `json:"opaqueData,omitempty"`
+	// Indication whether notifications to this endpoint are persistent or not
+	// +optional
+	Persistent bool `json:"persistent,omitempty"`
+	// Contains the endpoint spec of the topic
+	Endpoint TopicEndpointSpec `json:"endpoint"`
+}
+
+// TopicEndpointSpec contains exactly one of the endpoint specs of a Bucket Topic
+type TopicEndpointSpec struct {
+	// Spec of HTTP endpoint
+	// +optional
+	HTTP *HTTPEndpointSpec `json:"http,omitempty"`
+	// Spec of AMQP endpoint
+	// +optional
+	AMQP *AMQPEndpointSpec `json:"amqp,omitempty"`
+	// Spec of Kafka endpoint
+	// +optional
+	Kafka *KafkaEndpointSpec `json:"kafka,omitempty"`
+}
+
+// HTTPEndpointSpec represent the spec of an HTTP endpoint of a Bucket Topic
+type HTTPEndpointSpec struct {
+	// The URI of the HTTP endpoint to push notification to
+	// +kubebuilder:validation:MinLength=1
+	URI string `json:"uri"`
+	// Indicate whether the server certificate is validated by the client or not
+	// +optional
+	DisableVerifySSL bool `json:"disableVerifySSL,omitempty"`
+}
+
+// AMQPEndpointSpec represent the spec of an AMQP endpoint of a Bucket Topic
+type AMQPEndpointSpec struct {
+	// The URI of the AMQP endpoint to push notification to
+	// +kubebuilder:validation:MinLength=1
+	URI string `json:"endpoint"`
+	// Name of the exchange that is used to route messages based on topics
+	// +kubebuilder:validation:MinLength=1
+	Exchange string `json:"exchange"`
+	// Indicate whether the server certificate is validated by the client or not
+	// +optional
+	DisableVerifySSL bool `json:"disableVerifySSL,omitempty"`
+	// The ack level required for this topic (none/broker/routeable)
+	// +kubebuilder:validation:Enum=none;broker;routeable
+	// +kubebuilder:default=broker
+	// +optional
+	AckLevel string `json:"ackLevel,omitempty"`
+}
+
+// KafkaEndpointSpec represent the spec of a Kafka endpoint of a Bucket Topic
+type KafkaEndpointSpec struct {
+	// The URI of the Kafka endpoint to push notification to
+	// +kubebuilder:validation:MinLength=1
+	URI string `json:"endpoint"`
```

Comment:
I see `uri` in the design. Should this be `json:"uri"`, or should the design change?

Reply:
copy&paste bug